### PR TITLE
WIP: Update Argo for Swift 3

### PR DESCRIPTION
**Pull request by tonyd256**
This is a dumb update that simply went through the motions to get Argo
compiling in Xcode 8 beta 1. This should open up discussion on what
changes we should make for Swift 3.

### CODE REVIEW

File: Argo/Extensions/Dictionary.swift [hunk 1]
```diff
@@ -20,6 +20,6 @@ extension Dictionary {
   }
 }
 
-func <^> <T, U, V>(@noescape f: T -> U, x: [V: T]) -> [V: U] {
+func <^> <T, U, V>( f: @noescape (T) -> U, x: [V: T]) -> [V: U] {
```
Comment (gfontenot) [inline Argo/Extensions/Dictionary.swift]: Missing a `_` here?

File: Argo/Functions/flatReduce.swift [hunk 2]
```diff
@@ -17,7 +17,7 @@
   - returns: The result of iterating the combinator over every element of the
              sequence and flattening the result
 */
-public func flatReduce<S: SequenceType, U>(sequence: S, initial: U, @noescape combine: (U, S.Generator.Element) -> Decoded<U>) -> Decoded<U> {
+public func flatReduce<S: Sequence, U>(sequence: S, initial: U, combine: @noescape (U, S.Iterator.Element) -> Decoded<U>) -> Decoded<U> {
```
Comment (gfontenot) [inline Argo/Functions/flatReduce.swift]: Do we need to do `_ sequence` here?

File: Argo/Extensions/Dictionary.swift (hunk 1 above)
Comment (tonyd256) [inline Argo/Extensions/Dictionary.swift]: ah, no we don't need them on operators ... but the automatic conversion tool left that extra space in there so I need to fix that

File: Argo/Types/Decoded/Monad.swift [hunk 3]
```diff
@@ -28,7 +28,7 @@ public func >>- <T, U>(x: Decoded<T>, @noescape f: T -> Decoded<U>) -> Decoded<U
 
   - returns: A value of type `Decoded<U>`
 */
-public func -<< <T, U>(@noescape f: T -> Decoded<U>, x: Decoded<T>) -> Decoded<U> {
+public func -<< <T, U>( f: @noescape (T) -> Decoded<U>, x: Decoded<T>) -> Decoded<U> {
```
Comment (gfontenot) [inline Argo/Types/Decoded/Monad.swift]: There's a weird space before `f` here

File: Argo/Functions/curry.swift [hunk 4]
```diff
@@ -1,3 +1,3 @@
-func curry<T, U, V>(f : (T, U) -> V) -> T -> U -> V {
+func curry<T, U, V>(f : (T, U) -> V) -> (T) -> (U) -> V {
```
Comment (tonyd256) [inline Argo/Functions/curry.swift]: just realized we're not using this anymore ... will delete

File: Argo/Types/JSON.swift [hunk 5]
```diff
@@ -45,19 +45,20 @@ public extension JSON {
 }
 
 extension JSON: Decodable {
+  public typealias DecodedType = JSON
```
Comment (gfontenot) [inline Argo/Types/JSON.swift]: Why is this needed all of a sudden?

Comment (tonyd256) [inline Argo/Types/JSON.swift]: good catch .. it's not .. was messing around with a bug and forgot to remove

File: ArgoTests/JSON/JSONFileReader.swift [hunk 6]
```diff
@@ -1,13 +1,15 @@
 import Foundation
 
 func JSONFromFile(file: String) -> AnyObject? {
-  return NSBundle(forClass: JSONFileReader.self).pathForResource(file, ofType: "json")
-    .flatMap { NSData(contentsOfFile: $0) }
+  Bundle().pathForResource("user_without_key", ofType: "json")
```
Comment (gfontenot) [inline ArgoTests/JSON/JSONFileReader.swift]: What's going on here?

File: ArgoTests/JSON/JSONFileReader.swift [hunk 7]
```diff
@@ -1,13 +1,15 @@
...
+  return Bundle(for: JSONFileReader.self).pathForResource(file, ofType: "json")
+    .flatMap { URL(string: "file://\($0)") }
```
Comment (gfontenot) [inline ArgoTests/JSON/JSONFileReader.swift]: This is unfortunate

File: Argo/Functions/flatReduce.swift (hunk 2 above)
Comment (tonyd256) [inline Argo/Functions/flatReduce.swift]: maybe ... I intentionally left some function without `_`s to have a discussion about if it should since that's the way swift seems to be headed

**Comment by gfontenot** [comment]
Few comments, nothing major. I think this looks good in general. We _do_ need to update our enum naming to match Swift 3.0 IMO. It's going to take _forever_ for me to get used to lowercased enum cases but it's worth being consistent.

File: ArgoTests/JSON/JSONFileReader.swift (hunk 7 above)
Comment (tonyd256) [inline ArgoTests/JSON/JSONFileReader.swift]: ya looks like we lost that `NSData(contentsOfFile: String)` constructor ... and if we use the URL constructor then it needs a `file://` in front

File: ArgoTests/JSON/JSONFileReader.swift (hunk 6 above)
Comment (tonyd256) [inline ArgoTests/JSON/JSONFileReader.swift]: yup ... left in there by accident

File: Argo/Functions/flatReduce.swift (hunk 2 above)
Comment (gfontenot) [inline Argo/Functions/flatReduce.swift]: I think we should maintain the existing API here. I'd vote for ignoring the external label on `sequence`.

File: ArgoTests/JSON/JSONFileReader.swift [hunk 8]
```diff
@@ -1,13 +1,14 @@
 import Foundation
 
 func JSONFromFile(file: String) -> AnyObject? {
-  return NSBundle(forClass: JSONFileReader.self).pathForResource(file, ofType: "json")
-    .flatMap { NSData(contentsOfFile: $0) }
+  return Bundle(for: JSONFileReader.self).pathForResource(file, ofType: "json")
+    .flatMap { URL(string: "file://\($0)") }
```
Comment (gfontenot) [inline ArgoTests/JSON/JSONFileReader.swift]: We can use `URL(fileURLWithPath:)` here (pretty sure that's what it's called)

**Comment by gfontenot** [comment]
I'm OK with these changes. Was there anything specific you weren't sure about? I'd like to get this merged into `master` (along with a disclaimer about version compatibility) ASAP so that we can move forward with other Swift 3.0 changes that need to take place.
